raft/tracker/tracker.go

```diff
@@ -157,6 +157,11 @@ func (p *ProgressTracker) RemoveAny(id uint64) {
 		panic(fmt.Sprintf("peer %x is both voter and learner", id))
 	}
 
+	if okV1 && len(p.Voters[0]) == 1 {
+		// Never remove the last voter.
+		return
```

Comment:
Silently returning here seems like a potential problem. Should this be a panic just like the other invalid removals?

Reply:
What I am doing in a WIP (unpublished) is returning an error from `ApplyConfChange` which essentially allows the app to delegate config change checking to Raft itself. I'm doing this since it becomes more difficult to reason about what's allowed and what isn't when there are joint quorums in play. Additionally, the way conf changes are set up - the app passing in a "delta" - further complicates this, the app basically has to grab the current config, then compute what the final config would be, and decide whether to actually pass the delta to Raft - that's quite a bit of work that's easy to get wrong.

In CRDB, we'll additionally want to do this on the CRDB side too to avoid diverging our descriptor-encoded config and our Raft-encoded one (today we just don't check), but the code Raft uses to "check" the transition will be modular and so we can compute the final config on our end that way, check the result against the descriptor, and then feed it to Raft (knowing there won't be an error).

I wanted to send out a small PR making just this change, and I'd prefer not to panic since it's all the same (at least this way I can test this).

We also already have an unfortunate history of ignoring config changes hackily, for example those issued in `StartNode` and those with a `NodeID` of `None`. I will see about unifying all that down the road, though it's not my top priority.

Reply:
> I wanted to send out a small PR making just this change, and I'd prefer not to panic since it's all the same (at least this way I can test this).

My thinking is that since there are already (untested) panics here it makes sense to do the same for this new case, but I don't feel strongly about it (and you could still test it with RawNode by catching the panic). If you've got other cleanups coming this is fine.

Reply:
we probably should document this behavior at least if we do not panic here.